File: Year-2023/LeetCode-April-Challenge/LastStoneWeight.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq(stones.begin(), stones.end());
        while(pq.size() > 1){
            int firstTop = pq.top();
            pq.pop();
            int secondTop = pq.top();
            pq.pop();
            if(firstTop == secondTop) continue;
            else pq.push(firstTop - secondTop);
        }
        return pq.size() ? pq.top(): 0;
    }
};

int main(){
    vector<int> stones = {2,7,4,1,8,1};
    Solution *s = new Solution();
    cout<<s->lastStoneWeight(stones)<<endl;
    return 0;
}

File: Year-2023/LeetCode-April-Challenge/LastStoneWeight.java
import java.util.*;
class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for(int x: stones) pq.add(x);
        while(pq.size() > 1){
            int firstTop = pq.poll();
            int secondTop = pq.poll();
            if(firstTop == secondTop) continue;
            else pq.add(firstTop - secondTop);
        }
        return pq.size() != 0 ? pq.peek(): 0;
    }
}
public class LastStoneWeight {
    public static void main(String[] args) {
        int[] stones = {2,7,4,1,8,1};
        System.out.println(new Solution().lastStoneWeight(stones));
    }
}
